File: 4th semester/Systems for design and implementation/lab10/SocketsProject/common/src/main/java/Service/CastedSpellService.java
package Service;

import Model.CastedSpell;

import java.util.List;
import java.util.concurrent.Future;

public interface CastedSpellService {
    String ADD = "addCastedSpell";
    String DELETE = "deleteCastedSpell";
    String UPDATE = "updateCastedSpell";
    String GET = "getCastedSpells";

    Future<CastedSpell> addCastedSpell(long id, long id2, String description);
    Future<CastedSpell> updateCastedSpell(long id, long id2, String description);
    Future<CastedSpell> deleteCastedSpell(long id, long id2);
    Future<List<CastedSpell>> getCastedSpells();
}

File: 4th semester/Systems for design and implementation/lab11WebApp/src/app/casted-spell/shared/casted-spell.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map, Observable } from 'rxjs';
import { CastedSpell } from './castedSpell.model';

@Injectable({
  providedIn: 'root'
})
export class CastedSpellService {

  private url = "http://localhost:8080/api/casted_spells"

  constructor(private http: HttpClient) { }

  getCastedSpells(): Observable<CastedSpell[]> {
    return this.http.get<Array<CastedSpell>>(this.url).pipe(map(res => {return res['castedSpells']}));
  }

  getCastedSpell(id: number): Observable<CastedSpell> {
    return this.getCastedSpells().pipe(map(casted_spells => casted_spells.find(CastedSpell => CastedSpell.id === id)))
  }

  saveCastedSpell(CastedSpell: CastedSpell): Observable<CastedSpell> {
    console.log("save CastedSpell", CastedSpell)

    return this.http.post<CastedSpell>(this.url, CastedSpell)
  }

  updateCastedSpell(CastedSpell: CastedSpell): Observable<CastedSpell> {
    return this.http.put<CastedSpell>(this.url + "/" + CastedSpell.id, CastedSpell)
  }

  deleteCastedSpell(id: number): Observable<any> {
    return this.http.delete(this.url + "/" + id)
  }

  // filterCastedSpells(name: string) {
  //   return this.http.get<Array<CastedSpell>>(this.url + "/filter/" + name).pipe(map(res => {return res['castedSpells']}));
  // }

  // sortCastedSpells() {
  //   return this.http.get<Array<CastedSpell>>(this.url + "/sort").pipe(map(res => {return res['castedSpells']}));
  // }
}
